runtime/compiler/env/j9method.cpp

```diff
@@ -9412,3 +9413,50 @@ TR_J9ByteCodeIlGenerator::packReferenceChainOffsets(TR::Node *node, std::vector<
    return;
    }
 #endif
+
+bool
+TR_ResolvedJ9Method::isFieldQType(TR::Compilation *comp, int32_t cpIndex)
+   {
+   if (!TR::Compiler->om.areValueTypesEnabled() ||
+      (-1 == cpIndex))
+      return false;
+
+   J9ROMFieldRef *ref = (J9ROMFieldRef *) (&romCPBase()[cpIndex]);
+   J9ROMNameAndSignature *nameAndSignature = J9ROMFIELDREF_NAMEANDSIGNATURE(ref);
+   J9UTF8 *signature = J9ROMNAMEANDSIGNATURE_SIGNATURE(nameAndSignature);
+
+   J9JavaVM *vm = comp->fej9()->getJ9JITConfig()->javaVM;
+   return vm->internalVMFunctions->isNameOrSignatureQtype(signature);
+   }
+
+bool
+TR_ResolvedJ9Method::isFieldFlattened(TR::Compilation *comp, int32_t cpIndex)
+   {
+   if (!TR::Compiler->om.areValueTypesEnabled() ||
+      (-1 == cpIndex))
+      return false;
+
+   // Assumes the field is a QType.
+   int32_t fieldNameLength;
+   const char *fieldName = fieldNameChars(cpIndex, fieldNameLength);
+   int32_t sigLength;
+   const char *fieldSignature = fieldSignatureChars(cpIndex, sigLength);
+
+   bool isStatic;
+   TR_OpaqueClassBlock *containingClass = definingClassFromCPFieldRef(comp, cpIndex, isStatic);
+   TR_VMFieldsInfo fieldsInfo(comp, reinterpret_cast<J9Class*>(containingClass), 1, stackAlloc);
+   ListIterator<TR_VMField> iter(fieldsInfo.getFields());
+
+   for (TR_VMField *field = iter.getFirst(); field; field = iter.getNext())
```

Comment:
This seems inefficient, especially considering this method may be called for every field in an object (effectively making the traversal *O(n<sup>2</sup>)*). To make this more efficient, I was thinking that maybe this information could be stored in the TypeLayout/TypeLayoutEntry classes? TypeLayout also seems like a more natural place to ask this question, rather than the `TR_ResolvedJ9Method`.

Reply:
I found `TR_ResolvedJ9Method::definingClassFromCPFieldRef()` already retrieves `J9ROMFieldShape` from VM when it retrieves the defining class. There is no need to build `TR_VMFieldsInfo` for the containing class.